In [1]:
import seaborn as pd
import pandas as pd
import numpy as np

In [2]:
ad = pd.read_csv("Advertising.csv", usecols = [1, 2, 3, 4])
df = ad.copy()

In [3]:
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [45]:
X = df.drop("sales", axis = 1)  # bağımsız değişkenlerimiz

In [46]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [16]:
X.shape

(200, 3)

In [8]:
y = df["sales"]

In [10]:
y.head()

0    22.1
1    10.4
2     9.3
3    18.5
4    12.9
Name: sales, dtype: float64

In [17]:
y.shape

(200,)

In [9]:
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [18]:
X_train.shape

(160, 3)

In [19]:
y_train.shape

(160,)

In [20]:
X_test.shape

(40, 3)

In [21]:
y_test.shape

(40,)

### Görüldüğü üzere veri seti %80'e %20 olacak şekilde bölünmüş. test size = 0.2 bu yüzden yazıldı.

In [36]:
training = df.copy()

In [37]:
training.shape

(200, 4)

## Statsmodels

In [38]:
import statsmodels.api as sm

### R-squared'e bak. 'satış' değişkeninin değişkenliğinin yaklaşık %98'ini açıklıyoruz ki bu gerçek hayat için çok çok yüksek bir oran.

In [39]:
lm = sm.OLS(y_train, X_train)
model = lm.fit()
model.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.982
Model:                            OLS   Adj. R-squared:                  0.982
Method:                 Least Squares   F-statistic:                     2935.
Date:                Thu, 19 Sep 2019   Prob (F-statistic):          1.28e-137
Time:                        22:49:57   Log-Likelihood:                -336.65
No. Observations:                 160   AIC:                             679.3
Df Residuals:                     157   BIC:                             688.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0531      0.001     36.467      0.000       0.050       0.056
radio          0.2188      0.011     20.138      0.000       0.197       0.240
newspaper      0.0239      0.008      3.011      0.003       0.008       0.040
==============================================================================
Omnibus:                       11.405   Durbin-Watson:                   1.895
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               15.574
Skew:                          -0.432   Prob(JB):                     0.000415
Kurtosis:                       4.261   Cond. No.                         13.5
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### F-statistic, modelin anlamlılığını belirler.

### Adj. R-squared ve R-squared modelin anlamlı olup olmadığını belirler. 0.982 ise % 98 anlamlı olduğu anlamına gelir.

In [34]:
model.summary().tables[0]

<class 'statsmodels.iolib.table.SimpleTable'>

In [35]:
model.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

## scikit-learn model

In [40]:
from sklearn.linear_model import LinearRegression

In [41]:
lm = LinearRegression()
model = lm.fit(X_train, y_train)

In [42]:
model.intercept_  # sabit katsayımız.

2.979067338122629

In [44]:
model.coef_   # bağımsız değişkenlerin katsayıları

array([0.04472952, 0.18919505, 0.00276111])